File: SWB4/Client/OpenOffice/SWBOfficeCore/src/org/semanticwb/openoffice/OfficeApplication.java
/**  
* SemanticWebBuilder es una plataforma para el desarrollo de portales y aplicaciones de integración, 
* colaboración y conocimiento, que gracias al uso de tecnología semántica puede generar contextos de 
* información alrededor de algún tema de interés o bien integrar información y aplicaciones de diferentes 
* fuentes, donde a la información se le asigna un significado, de forma que pueda ser interpretada y 
* procesada por personas y/o sistemas, es una creación original del Fondo de Información y Documentación 
* para la Industria INFOTEC, cuyo registro se encuentra actualmente en trámite. 
* 
* INFOTEC pone a su disposición la herramienta SemanticWebBuilder a través de su licenciamiento abierto al público (‘open source’), 
* en virtud del cual, usted podrá usarlo en las mismas condiciones con que INFOTEC lo ha diseñado y puesto a su disposición; 
* aprender de él; distribuirlo a terceros; acceder a su código fuente y modificarlo, y combinarlo o enlazarlo con otro software, 
* todo ello de conformidad con los términos y condiciones de la LICENCIA ABIERTA AL PÚBLICO que otorga INFOTEC para la utilización 
* del SemanticWebBuilder 4.0. 
* 
* INFOTEC no otorga garantía sobre SemanticWebBuilder, de ninguna especie y naturaleza, ni implícita ni explícita, 
* siendo usted completamente responsable de la utilización que le dé y asumiendo la totalidad de los riesgos que puedan derivar 
* de la misma. 
* 
* Si usted tiene cualquier duda o comentario sobre SemanticWebBuilder, INFOTEC pone a su disposición la siguiente 
* dirección electrónica: 
*  http://www.semanticwebbuilder.org
**/ 
 
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.semanticwb.openoffice;

import java.awt.Frame;
import java.io.File;
import java.net.ConnectException;
import java.net.URI;
import java.net.URISyntaxException;
import java.text.SimpleDateFormat;
import java.util.List;
import javax.swing.JOptionPane;
import javax.swing.UIManager;
import org.netbeans.spi.wizard.Wizard;
import org.netbeans.spi.wizard.WizardPage;
import org.semanticwb.office.interfaces.WebPageInfo;
import org.semanticwb.openoffice.interfaces.IOpenOfficeApplication;
import org.semanticwb.openoffice.interfaces.IOpenOfficeDocument;
import org.semanticwb.openoffice.ui.dialogs.DialogAbout;
import org.semanticwb.openoffice.ui.dialogs.DialogChangePassword;
import org.semanticwb.openoffice.ui.dialogs.DialogDocumentsAuthorize;
import org.semanticwb.openoffice.ui.dialogs.DialogLogin;
import org.semanticwb.openoffice.ui.dialogs.DialogPreview;
import org.semanticwb.openoffice.ui.wizard.Search;
import org.semanticwb.openoffice.ui.wizard.SelectDirectory;
import org.semanticwb.openoffice.ui.wizard.SelectPage;
import org.semanticwb.openoffice.ui.wizard.SelectVersionToOpen;
import org.semanticwb.openoffice.ui.wizard.SelectWebPageID;
import org.semanticwb.openoffice.ui.wizard.TitleAndDescription;
import org.semanticwb.xmlrpc.HttpException;
import org.semanticwb.xmlrpc.XmlRpcProxyFactory;

/**
 *
 * @author victor.lorenzana
 */
public abstract class OfficeApplication
{

    public static SimpleDateFormat iso8601dateFormat = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
    private static final String DELETE_CONFIGURATION_ERROR = java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ERROR_TRATANDO_DE_OBTENER_EL_PROXY_SERVER,_FAVOR_DE_BORRA_EL_ARCHIVO_DE_CONFIGURACIÓN_UBICADO_EN:_");
    private static final String EMPTY_STRING = "";
    private static final String NL = "\r\n";
    private static final String HELP_URL = "http://www.webbuilder.org.mx";
    private static MenuListener menuListener;
    private static UserInfo userInfo = null;
    private static URI webAddress = null;
    private static IOpenOfficeDocument document;
    private static IOpenOfficeApplication application;


    static
    {       
        System.setProperty("wizard.sidebar.image", "org/semanticwb/openoffice/ui/icons/sidebar.png");
        System.setProperty("WizardDisplayer.default","org.semanticwb.openoffice.util.WBWizardDisplayerImpl");
        
        try
        {
            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
        }
        catch (Exception ue)
        {
            // No debe hacer nada
            System.out.println(ue.getMessage());
        }
    }    
    public static IOpenOfficeApplication getOfficeApplicationProxy() throws WBException
    {
        if (application == null)
        {
            application = XmlRpcProxyFactory.newInstance(IOpenOfficeApplication.class, OfficeApplication.getWebAddress());
            application.setUser(OfficeApplication.userInfo.getLogin());
            application.setPassword(OfficeApplication.userInfo.getPassword());
            String proxyServer = new Configuration().get(Configuration.PROXY_SERVER);
            String proxyPort = new Configuration().get(Configuration.PROXY_PORT);
            if (proxyServer == null)
            {
                proxyServer = EMPTY_STRING;
            }
            if (proxyPort == null)
            {
                proxyServer = EMPTY_STRING;
            }
            if (!proxyServer.equals(EMPTY_STRING) && !proxyPort.equals(EMPTY_STRING))
            {
                try
                {
                    application.setProxyAddress(new URI(proxyServer));
                    application.setProxyPort(Integer.parseInt(proxyPort));
                }
                catch (URISyntaxException e)
                {
                    String message = DELETE_CONFIGURATION_ERROR +" "+ new Configuration().getPath();
                    throw new WBException(message, e);
                }
                catch (NumberFormatException e)
                {
                    String message = DELETE_CONFIGURATION_ERROR+" " + new Configuration().getPath();
                    throw new WBException(message, e);
                }
            }
            try
            {
                if (!application.isValidVersion(IOpenOfficeApplication.version))
                {
                    throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("LA_VERSIÓN_ENTRE_LA_APLICACIÓN_DE_PUBLICACIÓN_Y_EL_SITIO_NO_ES_COMPATIBLE"));
                }
            }
            catch (ConnectException e)
            {
                JOptionPane.showMessageDialog(null, java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_CONECTAR_AL_SERVIDORDETALLE:_")+" "+e.getLocalizedMessage(), java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ERROR_DE_ACCESO"), JOptionPane.OK_OPTION |
                            JOptionPane.ERROR_MESSAGE);
                throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_CONECTAR_AL_SERVIDORDETALLE:_")+" "+e.getLocalizedMessage(), e);
            }
            catch (HttpException e)
            {
                if (e.getCode() == 403 || e.getCode()==404)
                {
                    JOptionPane.showMessageDialog(null, e.getLocalizedMessage(), java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ERROR_DE_ACCESO"), JOptionPane.OK_OPTION |
                            JOptionPane.ERROR_MESSAGE);
                    throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_VALIDAR_LA_COMPATIBILIDAD_DE_VERSIONES") + e.getMessage(), e);
                }
                else
                {
                    throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_VALIDAR_LA_COMPATIBILIDAD_DE_VERSIONES") + e.getMessage(), e);
                }
            }
            catch (Exception e)
            {
                throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ERROR_AL_TRATAR_DE_VERIFICAR_COMPATIBILIDAD_DE_VERSIONES_ENTRE_EL_PUBLICADOR_Y_EL_SERVIDOR"), e);

            }
        }
        return application;
    }

    private static void setProxy() throws WBException
    {
        String proxyServer = new Configuration().get(Configuration.PROXY_SERVER);
        String proxyPort = new Configuration().get(Configuration.PROXY_PORT);
        if (proxyServer == null)
        {
            proxyServer = EMPTY_STRING;
        }
        if (proxyPort == null)
        {
            proxyServer = EMPTY_STRING;
        }
        if (!proxyServer.equals(EMPTY_STRING) && !proxyPort.equals(EMPTY_STRING))
        {
            try
            {
                document.setProxyAddress(new URI(proxyServer));
                document.setProxyPort(Integer.parseInt(proxyPort));
            }
            catch (URISyntaxException e)
            {
                String message = DELETE_CONFIGURATION_ERROR +" "+ new Configuration().getPath();
                throw new WBException(message, e);
            }
            catch (NumberFormatException e)
            {
                String message = DELETE_CONFIGURATION_ERROR+" " + new Configuration().getPath();
                throw new WBException(message, e);
            }
        }
    }

    public static IOpenOfficeDocument getOfficeDocumentProxy() throws WBException
    {
        if (document == null)
        {
            try
            {
                if (getOfficeApplicationProxy().isValidVersion(IOpenOfficeApplication.version))
                {
                    document = XmlRpcProxyFactory.newInstance(IOpenOfficeDocument.class, OfficeApplication.getWebAddress());
                    document.setUser(OfficeApplication.userInfo.getLogin());
                    document.setPassword(OfficeApplication.userInfo.getPassword());
                    setProxy();
                }
            }
            catch (ConnectException e)
            {
                JOptionPane.showMessageDialog(null, java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_CONECTAR_AL_SERVIDORDETALLE:_")+" "+e.getLocalizedMessage(), java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ERROR_DE_ACCESO"), JOptionPane.OK_OPTION |
                            JOptionPane.ERROR_MESSAGE);
                throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_CONECTAR_AL_SERVIDORDETALLE:_")+" "+e.getLocalizedMessage(), e);
            }
            catch (HttpException e)
            {
                if (e.getCode() == 403)
                {
                    JOptionPane.showMessageDialog(null, e.getLocalizedMessage(), java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ERROR_DE_ACCESO"), JOptionPane.OK_OPTION |
                            JOptionPane.ERROR_MESSAGE);
                    throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_VALIDAR_LA_COMPATIBILIDAD_DE_VERSIONES") + e.getMessage(), e);
                }
                else
                {
                    throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_VALIDAR_LA_COMPATIBILIDAD_DE_VERSIONES") + e.getMessage(), e);
                }
            }
            catch (Exception e)
            {
                throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("NO_SE_PUEDE_VALIDAR_LA_COMPATIBILIDAD_DE_VERSIONES") + e.getMessage(), e);
            }
        }
        return document;
    }

    public void setMenuListener(MenuListener menuListener)
    {
        OfficeApplication.menuListener = menuListener;
    }

    protected OfficeApplication()
    {
    }

    /*private void verifyVersion()
    {
    if ( OfficeApplication.tryLogin() )
    {
    IOpenOfficeApplication officeApplication = XmlRpcProxyFactory.newInstance(IOpenOfficeApplication.class, webAddress);
    try
    {
    if ( officeApplication.isValidVersion(IOpenOfficeApplication.version) && menuListener != null )
    {
    menuListener.onLogin();
    }
    }
    catch ( Exception e )
    {
    JOptionPane.showMessageDialog(null, e.getMessage(), "Verificación de versión de publicación", JOptionPane.ERROR_MESSAGE);
    }
    }
    }*/
    /**
     * Opens a document in a file path
     * @param file Path for the file
     * @return OfficeDocument opened
     * @throws org.semanticwb.WBException If the document can not be opened
     */
    public abstract OfficeDocument open(File file) throws WBException;

    /**
     * Returns all documents opened
     * @return List of documents
     * @throws org.semanticwb.WBException In case that there was an error
     */
    public abstract List<OfficeDocument> getDocuments() throws WBException;

    public static final void changePassword()
    {
        if (OfficeApplication.tryLogin())
        {
            DialogChangePassword dialog = new DialogChangePassword();
            dialog.setLocationRelativeTo(null);
            dialog.setVisible(true);
            if (!dialog.isCanceled())
            {
                try
                {
                    getOfficeApplicationProxy().changePassword(dialog.getNewPassword());
                    userInfo.changePassword(dialog.getNewPassword());
                }
                catch (Exception e)
                {
                    JOptionPane.showMessageDialog(null, e.getLocalizedMessage(), java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("CAMBIAR_CONTRASEÑA"), JOptionPane.ERROR);
                }
            }
        }
    }

    public static final void createPage()
    {
        if (OfficeApplication.tryLogin())
        {
            CreatePageResultProducer resultProducer = new CreatePageResultProducer();            
            WizardPage[] clazz = new WizardPage[]
            {
                new SelectPage(null), new TitleAndDescription(false), new SelectWebPageID()
            };
            Wizard wiz = WizardPage.createWizard(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ASISTENTE_DE_CREACIÓN_DE_PÁGINA"), clazz, resultProducer);
            wiz.show();
        }
    }

    public static final void createPage(WebPageInfo parent)
    {
        CreatePageResultProducer resultProducer = new CreatePageResultProducer(parent);
        WizardPage[] clazz = new WizardPage[]
        {
            new TitleAndDescription(false), new SelectWebPageID(parent)
        };
        Wizard wiz = WizardPage.createWizard(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ASISTENTE_DE_CREACIÓN_DE_PÁGINA"), clazz, resultProducer);
        wiz.show();
    }

    public final void open(DocumentType type)
    {
        if (OfficeApplication.tryLogin())
        {
            OpenResultProducer resultProducer = new OpenResultProducer(this);
            WizardPage[] clazz = new WizardPage[]
            {
                new Search(type), new SelectVersionToOpen(type.toString().toLowerCase()), new SelectDirectory()
            };
            Wizard wiz = WizardPage.createWizard(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("ASISTENTE_DE_APERTURA_DE_CONTENIDO"), clazz, resultProducer);
            wiz.show();
        }
    }

    public static final void showAbout()
    {
        DialogAbout dialog = new DialogAbout();
        dialog.setVisible(true);
    }

    public static final void showHelp()
    {
        DialogPreview.showInBrowser(HELP_URL, new Frame());
    }

    public static URI getWebAddress() throws WBException
    {

        if (webAddress == null)
        {
            if (!tryLogin())
            {
                throw new WBException(java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("EL_USUARIO_NO_PUEDE_ACCEDER"));
            }
        }
        return webAddress;
    }

    static String setupDocument(String workspace, String contentId) throws Exception
    {
        String contentIdToReturn = null;
        if (contentId != null && !contentId.trim().equals(EMPTY_STRING) && OfficeApplication.tryLogin())
        {
            document = getOfficeDocumentProxy();
            try
            {
                if (document.exists(workspace, contentId))
                {
                    contentIdToReturn = contentId;
                }
                else
                {
                    // el contenido no existe en el sitio pero indica que ya tiene un identificador
                    JOptionPane.showMessageDialog(null, java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("EL_CONTENIDO_PARACE_HABERSE_PUBLICADO_EN_UN_SITIO_WEB.")+ NL +java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("AL_SITIO_DONDE_SE_ESTÁ_INTENTANDO_CONECTAR,_INDICA_QUE_ESTE_CONTENIDO_NO_EXISTE.")+ NL +java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("SI_DESEA_CONTINUAR_SE_PERDERÁ_ESTA_INFORMACIÓN,_DE_LO_CONTRARIO,_CIERRE_ESTE_DOCUMENTO."), java.util.ResourceBundle.getBundle("org/semanticwb/openoffice/OfficeApplication").getString("VERIFICACIÓN_DE_CONTENIDO_EN_SITIO_WEB"), JOptionPane.WARNING_MESSAGE);
                }
            }
            catch (NumberFormatException nfe)
            {
                ErrorLog.log(nfe);
            }

        }
        return contentIdToReturn;
    }
    /*public static MenuListener getMenuListener()
    {
    return menuListener;
    }*/

    private final static boolean logOn()
    {
        boolean logOn = false;
        DialogLogin frmlogin = new DialogLogin();        
        
        frmlogin.setVisible(true);
        if (!frmlogin.isCanceled())
        {
            String login = frmlogin.getLogin();
            String password = frmlogin.getPassword();
            userInfo = new UserInfo(password, login);
            webAddress = frmlogin.getWebAddress();
            logOn = true;
        }
        else
        {
            logOff();
        }
        return logOn;
    }

    public static boolean tryLogin()
    {
        boolean tryLogin = false;
        if (userInfo == null || webAddress == null)
        {
            logOn();            
            
            if (userInfo == null || webAddress == null)
            {
                logOff();
                tryLogin = false;
            }
            else
            {                
                try
                {
                    OfficeApplication.getOfficeApplicationProxy().isValidVersion(IOpenOfficeApplication.version);
                    tryLogin = true;
                    ConfigurationListURI configurationListURI = new ConfigurationListURI();
                    configurationListURI.addUserConfiguration(webAddress,userInfo.getLogin());
                }
                catch (Exception e)
                {                    
                    tryLogin = false;
                    logOff();
                }
            }
        }
        else
        {
            tryLogin = true;
        }
        return tryLogin;
    }

    public static final boolean isLogged()
    {
        boolean isLogged = false;
        if (userInfo != null)
        {
            isLogged = true;
        }
        return isLogged;
    }
    public static final void openSession()
    {
        tryLogin();
    }
    public static final void closeSession()
    {
        logOff();
    }
    public static final void showContentsToAuthorize()
    {
        if(tryLogin())
        {
            DialogDocumentsAuthorize dialogDocumentsAtuhorize=new DialogDocumentsAuthorize();
            dialogDocumentsAtuhorize.setVisible(true);
        }
    }

    public final static void logOff()
    {
        userInfo = null;
        webAddress = null;
        application=null;
        if (menuListener != null)
        {
            menuListener.onLogout();
        }
    }
}

File: SWB4/Client/Microsoft Office/INFOTEC WebBuilder 4/WBOffice4/OfficeApplication.cs
/**  
* SemanticWebBuilder es una plataforma para el desarrollo de portales y aplicaciones de integración, 
* colaboración y conocimiento, que gracias al uso de tecnología semántica puede generar contextos de 
* información alrededor de algún tema de interés o bien integrar información y aplicaciones de diferentes 
* fuentes, donde a la información se le asigna un significado, de forma que pueda ser interpretada y 
* procesada por personas y/o sistemas, es una creación original del Fondo de Información y Documentación 
* para la Industria INFOTEC, cuyo registro se encuentra actualmente en trámite. 
* 
* INFOTEC pone a su disposición la herramienta SemanticWebBuilder a través de su licenciamiento abierto al público (‘open source’), 
* en virtud del cual, usted podrá usarlo en las mismas condiciones con que INFOTEC lo ha diseñado y puesto a su disposición; 
* aprender de él; distribuirlo a terceros; acceder a su código fuente y modificarlo, y combinarlo o enlazarlo con otro software, 
* todo ello de conformidad con los términos y condiciones de la LICENCIA ABIERTA AL PÚBLICO que otorga INFOTEC para la utilización 
* del SemanticWebBuilder 4.0. 
* 
* INFOTEC no otorga garantía sobre SemanticWebBuilder, de ninguna especie y naturaleza, ni implícita ni explícita, 
* siendo usted completamente responsable de la utilización que le dé y asumiendo la totalidad de los riesgos que puedan derivar 
* de la misma. 
* 
* Si usted tiene cualquier duda o comentario sobre SemanticWebBuilder, INFOTEC pone a su disposición la siguiente 
* dirección electrónica: 
*  http://www.webbuilder.org.mx 
**/ 
 
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using WBOffice4.Steps;
using WBOffice4.Forms;
using System.Windows.Forms;
using XmlRpcLibrary;
using WBOffice4.Interfaces;
using System.Globalization;
using System.Diagnostics;
using System.Net;
using WBOffice4.Utils;
using System.Threading;
namespace WBOffice4
{

    
    public abstract class OfficeApplication
    {
        
        public static readonly String iso8601dateFormat = "{0:dd/MM/yyyy HH:mm:ss}";
        public static readonly TraceEventLogListener listener = new TraceEventLogListener();
        private static UserInfo userInfo;
        private static Uri webAddress;
        private static IMenuListener _MenuListener;
        private static IOfficeDocument officedocument;
        private static IOfficeApplication officeApplication;
        public static readonly double m_version = 4.100;
        static OfficeApplication()
        {            
            try
            {
                Thread.CurrentThread.CurrentUICulture = new CultureInfo("es");
            }
            catch (Exception e)
            {
                OfficeApplication.WriteError(e);
                //MessageBox.Show(e.Message + "\r\n" + e.StackTrace);
            }
            Debug.Listeners.Add(listener);
            Trace.Listeners.Add(listener);
        }
        protected OfficeApplication()
        {
                      
        }
        public static void WriteError(Exception e)
        {            
            listener.WriteError(e);
        }
        internal static IOfficeDocument OfficeDocumentProxy
        {
            get
            {
                if (officedocument == null)
                {
                    if (OfficeApplicationProxy.isValidVersion(m_version))
                    {
                        if (_MenuListener != null)
                        {
                            _MenuListener.LogOn();
                        }
                        officedocument = XmlRpcProxyFactory.Create<IOfficeDocument>();
                        officedocument.WebAddress = GetWebAddress();
                        SWBConfiguration configuration = new SWBConfiguration();
                        officedocument.Credentials = new NetworkCredential(userInfo.Login, userInfo.Password);
                        if (configuration.UsesProxy)
                        {
                            officedocument.ProxyPort = int.Parse(configuration.ProxyPort,CultureInfo.InvariantCulture);
                            officedocument.ProxyServer = new Uri(configuration.ProxyServer);
                        }
                    }
                }
                return officedocument;
            }

        }
        internal static IOfficeApplication OfficeApplicationProxy
        {
            get
            {
                if (officeApplication == null)
                {
                    officeApplication = XmlRpcProxyFactory.Create<IOfficeApplication>();
                    officeApplication.WebAddress = GetWebAddress();
                    SWBConfiguration configuration = new SWBConfiguration();
                    officeApplication.Credentials = new NetworkCredential(userInfo.Login, userInfo.Password);
                    if (configuration.UsesProxy)
                    {
                        officeApplication.ProxyPort = int.Parse(configuration.ProxyPort,CultureInfo.InvariantCulture);
                        officeApplication.ProxyServer = new Uri(configuration.ProxyServer);
                    }
                    if (!officeApplication.isValidVersion(m_version))
                    {
                        throw new WBAlertException("La versión entre el publicador y el servidor no son compatibles");
                    }
                    else
                    {
                        RtlAwareMessageBox.Show("Su sesión ha sido iniciada", "Iniciar sessión", MessageBoxButtons.OK, MessageBoxIcon.Information);
                    }
                }
                return officeApplication;
            }
        }
        public static IMenuListener MenuListener
        {
            get
            {
                return _MenuListener;
            }
            set
            {
                _MenuListener = value;
            }
        }       

        /// <summary>
        /// Gets all the documents opened in the application
        /// </summary>
        protected internal abstract ICollection<OfficeDocument> Documents
        {
            get;
            
        }
        protected internal abstract string Version
        {
            get;
        }

        protected internal abstract OfficeDocument Open(FileInfo file);

        protected internal abstract OfficeDocument Open(FileInfo file,String contentid,String rep);

        /// <param name="obj">Clase <see cref="T:System.Object"/> que se va a comparar con la clase <see cref="T:System.Object"/> actual.</param>
        /// <returns>Es true si el objeto <see cref="T:System.Object"/> especificado es igual al objeto <see cref="T:System.Object"/> actual; en caso contrario, es false.</returns>
        public override bool Equals(object obj)
        {
            OfficeApplication officeApplication=obj as OfficeApplication;
            if (officeApplication!=null)
            {                
                return this.Version.Equals(officeApplication.Version);
            }
            return false;
        }

        /// <returns>Código hash para la clase <see cref="T:System.Object"/> actual.</returns>
        public override int GetHashCode()
        {
            return this.Version.GetHashCode();
        }

        /// <returns>Una clase <see cref="T:System.String"/> que representa la clase <see cref="T:System.Object"/> actual.</returns>
        public override string ToString()
        {
            return this.Version;
        }
        public void Open(DocumentType documentType)
        {
            if (OfficeApplication.TryLogOn())
            {
                FormOpenContent frm = new FormOpenContent(this, documentType);
                frm.ShowDialog();
                frm.open();
            }
        }
        public static void ChangePassword()
        {
            if (TryLogOn())
            {
                FormChangePassword form = new FormChangePassword();
                DialogResult result=form.ShowDialog();
                if (result == DialogResult.OK)
                {
                    OfficeApplicationProxy.changePassword(form.NewPassword);
                    userInfo.ChangePassword(form.NewPassword);
                }
            }
        }
        internal static void LogOff()
        {
            officeApplication = null;
            officedocument = null;
            userInfo = null;
            webAddress = null;            
        }
        public static void CloseSession()
        {
            userInfo = null;
            webAddress = null;
            if (MenuListener != null)
            {
                MenuListener.LogOff();
                LogOff();
            }

        }
        internal protected static string SetupDocument(String repositoryName,String contentId)
        {            
            if (!String.IsNullOrEmpty(contentId) && OfficeApplication.TryLogOn())
            {
                IOfficeDocument document=OfficeApplication.OfficeDocumentProxy;
                try
                {
                    if (document.exists(repositoryName,contentId))
                    {
                        return contentId;
                    }
                    else
                    {
                        RtlAwareMessageBox.Show("El contenido parace haberse publicado en un sitio web.\r\nAl sitio donde se está intentando conectar, indica que este contenido no existe.\r\nSi desea continuar se perderá esta información, de lo contrario, cierre este documento.", "Verificación de contenido en sitio web", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    }
                }
                catch (FormatException fe)
                {
                    OfficeApplication.WriteError(fe);
                }
                catch (ApplicationException ap)
                {
                    OfficeApplication.WriteError(ap);
                }                
            }
            return null;
        }        
        internal static Uri GetWebAddress()
        {            
            if (webAddress == null)
            {
                if (TryLogOn())
                {
                    return webAddress;
                }
                else
                {
                    throw new WBAlertException("The user can be logged");
                }
            }
            return webAddress;             
        }
        public static void InitSession()
        {
            TryLogOn();            
        }
        internal static bool TryLogOn()
        {
            bool tryLogOn = false;
            if (userInfo == null || webAddress == null)
            {
                LogOn();
                if (userInfo == null || webAddress == null)
                {
                    LogOff();
                    tryLogOn= false;
                }
                else
                {
                    MenuListener.LogOn();
                    tryLogOn=true;
                }
            }
            else
            {
                tryLogOn=true;
            }
            if (tryLogOn)
            {
                try
                {
                    OfficeApplicationProxy.isValidVersion(m_version);
                    if (MenuListener != null)
                    {
                        MenuListener.LogOn();
                    }
                    URIConfigurationList uRIConfigurationList = new URIConfigurationList();
                    String uri=OfficeApplicationProxy.WebAddress.ToString();
                    if (uri.EndsWith("gtw"))
                    {
                        uri = uri.Substring(0, uri.Length - 3);
                    }
                    uRIConfigurationList.Add(OfficeApplicationProxy.Credentials.UserName, new Uri(uri));
                    MenuListener.LogOn();                    
                }
                catch (HttpException e)
                {
                    MenuListener.LogOff();
                    tryLogOn = false;
                    if (e.Code == HttpStatusCode.NotFound)
                    {                        
                        RtlAwareMessageBox.Show(null, "No se puede conectar a la dirección web, o la dirección es incorrecta o no se tiene habilitada la publicación de contenidos", "Iniciar sessión", MessageBoxButtons.OK, MessageBoxIcon.Error);
                        
                    }
                    else if (e.Code == HttpStatusCode.Forbidden)
                    {
                        RtlAwareMessageBox.Show(null, "Su clave o contraseña es incorrecta", "Iniciar sessión", MessageBoxButtons.OK, MessageBoxIcon.Error);                        
                    }
                    else
                    {
                        RtlAwareMessageBox.Show(null, e.Message, "Iniciar sessión", MessageBoxButtons.OK, MessageBoxIcon.Error, MessageBoxDefaultButton.Button1);
                        
                    }
                    OfficeApplication.LogOff();
                }
                catch (WebException e)
                {
                    MenuListener.LogOff();
                    tryLogOn = false;
                    if (e.Status == WebExceptionStatus.ConnectFailure)
                    {
                        RtlAwareMessageBox.Show(null, "El sitio web con el que intenta trabajar, se encuentra apagado o no se puede acceder al mismo.", "Iniciar sessión", MessageBoxButtons.OK, MessageBoxIcon.Error);                        
                    }
                    else
                    {
                        RtlAwareMessageBox.Show(null, e.Message, "Iniciar sessión", MessageBoxButtons.OK, MessageBoxIcon.Error);                        
                    }
                    OfficeApplication.LogOff();
                }               
                catch (Exception e)
                {
                    OfficeApplication.WriteError(e);
                    MenuListener.LogOff();
                    RtlAwareMessageBox.Show(null, e.Message, "Iniciar sessión", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    tryLogOn = false;
                    OfficeApplication.LogOff();
                }                
            }
            return tryLogOn;
        }
        private static void LogOn()
        {
            FormLogin login = new FormLogin();
            DialogResult result=login.ShowDialog();
            if (result == DialogResult.OK)
            {
                userInfo = new UserInfo(login.User, login.Password);
                webAddress = login.WebAddress;
            }
            else
            {
                LogOff();
            }
        }
        public static void ShowHelp()
        {
            try
            {
                System.Diagnostics.Process.Start("explorer.exe", "\"http://www.semanticwebbuilder.org.mx/OS/wb/COSWB/documentation\"");

            }
            catch (Exception err)
            {
                OfficeApplication.WriteError(err);                
            }

        }
        public static void ShowAbout()
        {
            AboutBox frmAbout = new AboutBox();
            frmAbout.ShowDialog();
        }
        public static void CreatePage()
        {
            if (TryLogOn())
            {
                FormCreatePage form = new FormCreatePage();
                form.ShowDialog();                
            }
        }
        public static void ShowDocumentsToAuthorize()
        {
            if (TryLogOn())
            {
                FormDocumentsToAuthorize form = new FormDocumentsToAuthorize();
                form.ShowDialog();
            }
        }
        
    }
}
